examples/linkedlist.py

```diff
@@ -0,0 +1,65 @@
+from __future__ import print_function, absolute_import
+from numba.utils import OrderedDict
+import numpy as np
+from numba import njit
+from numba.jitclass import jitclass
+from numba import int32, deferred_type, optional
+from numba.runtime import rtsys
+
+node_type = deferred_type()
+
+spec = OrderedDict()
+spec['data'] = int32
+spec['next'] = optional(node_type)
+
+
+@jitclass(spec)
+class LinkedNode(object):
+    def __init__(self, data, next):
+        self.data = data
+        self.next = next
+
+    def prepend(self, data):
+        return LinkedNode(data, self)
+
+
+@njit
+def make_linked_node(data):
+    return LinkedNode(data, None)
+
+
+node_type.define(LinkedNode.class_type.instance_type)
```

Comment:
While the recursive data structure examples are pleasant, I'm a bit skeptical about the deffered type design. The deferred type ends up leaking in all aspects of typing and code generation, and we have to special-case some operations for it (as is already the case with Optional).

Perhaps all deferred references could be resolved at type instantiation, so for example instead of writing:
```node_type.define(LinkedNode.class_type.instance_type)```
you would write :
``node_type = node_type.resolve(LinkedNode.class_type.instance_type)```
?